src/cryptonote_protocol/cryptonote_protocol_handler.inl

```diff
@@ -361,6 +366,264 @@ namespace cryptonote
   template<class t_core>
+  int t_cryptonote_protocol_handler<t_core>::handle_notify_new_fluffy_block(int command, NOTIFY_NEW_FLUFFY_BLOCK::request& arg, cryptonote_connection_context& context)
+  {
+    LOG_PRINT_CCONTEXT_L2("NOTIFY_NEW_FLUFFY_BLOCK (hop " << arg.hop << ")");
+    if(context.m_state != cryptonote_connection_context::state_normal)
+      return 1;
+    
+    m_core.pause_mine();
+      
+    block new_block;
+    transaction miner_tx;
+    if(parse_and_validate_block_from_blob(arg.b.block, new_block))
+    {
+      std::list<blobdata> have_tx;
+      
+      // Instead of requesting missing transactions by hash like BTC, 
+      // we do it by index (thanks to a suggestion from moneromooo) because
+      // we're way cooler .. and also because they're smaller than hashes.
+      // 
+      // Also, remember to pepper some whitespace changes around to bother
+      // moneromooo ... only because I <3 him. 
+      std::list<size_t> need_tx_indices;
+        
+      transaction tx;
+      crypto::hash tx_hash;
+      std::unordered_map<crypto::hash, blobdata> tx_hash_to_blob;        
+      BOOST_FOREACH(auto& tx_blob, arg.b.txs)
+      {
+        if(parse_and_validate_tx_from_blob(tx_blob, tx))
+        {
+          try
+          {
+            if(!get_transaction_hash(tx, tx_hash))
+            {
+              LOG_PRINT_CCONTEXT_L1("get_transaction_hash failed");
+              m_p2p->drop_connection(context);
+              m_core.resume_mine();
+              return 1;
+            }
+          }
+          catch(...)
+          {
+            LOG_PRINT_CCONTEXT_L1("get_transaction_hash failed");
+            m_p2p->drop_connection(context);
+            m_core.resume_mine();
+            return 1;
+          }
+          
+          // we might already have the tx that the peer
+          // sent in our pool, so don't verify again..
+          if(!m_core.get_pool_transaction(tx_hash, tx))
+          {
+            cryptonote::tx_verification_context tvc = AUTO_VAL_INIT(tvc);                        
+            if(!m_core.handle_incoming_tx(tx_blob, tvc, true, true) || tvc.m_verifivation_failed)
+            {
+              LOG_PRINT_CCONTEXT_L1("Block verification failed: transaction verification failed, dropping connection");
+              m_p2p->drop_connection(context);
+              m_core.resume_mine();
+              return 1;
+            }
+            
+            // Should only not be added to pool if verification failed, but
+            // maybe in the future could not be added for other reasons 
+            // according to monero-moo so keep track of these separately ..
+            //
+            // this is probably the wrong thing to do, but for now the idea
+            // would be to pass these back/forth on the missing tx request,
+            // right now they're just being lost at the end of this function call
+            if(!tvc.m_added_to_pool) 
+              tx_hash_to_blob[tx_hash] = tx_blob;
```

Comment:
This will need some changes to the pool stuff, so these can be kept, at least temporarily, even if in the future we limit the pool. This is some future change though.